src/lib/models/ConfigModel.ts

```diff
@@ -1,172 +1,194 @@
-    #                           Default: ['Recommended']
-    #
-    # [Example usage]:
-    # engines:
-    #   regex:
-    #     custom_rules:
-    #       NoTodoComments:
-    #         regex: /\\/\\/[ \\t]*TODO/gi
-    #         file_extensions: [".cls", ".trigger"]
-    #         description: "Prevents TODO comments from being in apex code."
-    #         violation_message: "A comment with a TODO statement was found. Please remove TODO statements from your apex code."
-    #         severity: "Info"
-    #         tags: ['TechDebt']
-    custom_rules: {} # Remove this empty object {} when you are ready to define your first custom rule
-
-  # Custom configuration settings for the 'retire-js' engine
-  # See <LINK_COMING_SOON> to learn more about these settings.
-  retire-js:
-
-    # Whether to turn off the 'retire-js' engine so that it is not included when running Code Analyzer commands
-    disable_engine: false`;
+function getDefaultRule(defaultState: ConfigState, engineName: string, ruleName: string): Rule|null {
+	try {
+		return defaultState.rules.getRule(engineName, ruleName);
+	} catch (e) {
+		return null;
+	}
+}
+
+function toYamlRule(userRule: Rule, defaultRule: Rule|null, styled: boolean): string {
+	const ruleName: string = userRule.getName();
+	const userSeverity: SeverityLevel = userRule.getSeverityLevel();
+	const userTags: string[] = userRule.getTags();
+
+	let severityYaml = yaml.dump({severity: userSeverity});
+	let tagsYaml = yaml.dump({tags: userTags});
+
+	if (defaultRule != null) {
+		const defaultSeverity: SeverityLevel = defaultRule.getSeverityLevel();
+		const defaultTagsJson: string = JSON.stringify(defaultRule.getTags());
+
+		if (userSeverity !== defaultSeverity) {
+			const comment = getMessage(BundleName.ConfigModel, 'template.modified-from', [defaultSeverity]);
+			severityYaml = severityYaml.replace('\n', ` ${toYamlComment(comment, styled)}\n`);
+		}
+
+		if (JSON.stringify(userTags) !== defaultTagsJson) {
+			const comment = getMessage(BundleName.ConfigModel, 'template.modified-from', [defaultTagsJson]);
+			// The YAML spec requires a trailing newline, so we know that there's at least one newline somewhere in the
+			// string. If we inject a comment before the first newline we encounter, then it will look clean.
+			tagsYaml = tagsYaml.replace('\n', ` ${toYamlComment(comment, styled)}\n`);
+		}
+	}
+	return `"${ruleName}":\n${indent(severityYaml, 2)}${indent(tagsYaml, 2)}`;
+}
+
+function toYamlEngines(userState: ConfigState, styled: boolean): string {
+	if (userState.rules.getCount() === 0) {
+		const comment = getMessage(BundleName.ConfigModel, 'template.yaml.no-engines-selected');
+		return `engines: {} ${toYamlComment(comment, styled)}`;
+	}
+
+	let results: string = 'engines:\n'
+
+	for (const engineName of userState.core.getEngineNames()) {
+		if (!userState.rules.getEngineNames().includes(engineName)) {
+			continue;
+		}
```

Comment:
This is where I recommend looping over just the relevant engines (which I discussed how to derive in a previous comment).